Fix running NixOS tests on darwin

roberth opened the pull request:
- Fixes https://github.com/NixOS/nixpkgs/pull/303597#issuecomment-2864151691

- :bulb:  Review with[ whitespace changes hidden](https://github.com/NixOS/nixpkgs/pull/405599/files?w=1)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
Revert "nixos/lib/testing: avoid generating darwin VM tests"

This reverts commit 72155225aa62c6ac9bc67a1253079f140045511f.

roberth committed:
nixos/all-tests: Set hydraPlatforms

This only sets hydraPlatforms for the in-tree tests.
See the preceding revert commit which reverts the previous solution.

Co-Authored-By: Vladimír Čunát <v@cunat.cz>

roberth committed:
nixos/testing: Add meta.hydraPlatforms

roberth committed:
nixos/testing/meta.nix: lib.mkOption -> mkOption

roberth commented:
@ofborg build nixosTests.simple

Fully expecting that the build fails because of a lack of virtualization on ofborg Darwin hosts. This is for checking the eval only.

vcunat commented:
I think that this will make (some of) those jobs reappear on Hydra.  Random test case:
```
$ nix eval -f pkgs/top-level/release.nix tests.testers.nixosTest-example.x86_64-darwin.outPath
"/nix/store/qnkk51l596x4sa9fip2m3mwrfrbinmhw-vm-test-run-nixosTest-test"
```

roberth committed:
nixos/testing: Drop darwin from default hydraPlatforms

Litmus test: the following attribute disappears

    nix eval -f pkgs/top-level/release.nix tests.testers.nixosTest-example.x86_64-darwin.outPath

roberth committed:
nixos/all-tests: Rely on general hydraPlatforms override

It is documented to be about `hydra.nixos.org`, so this is actually
acceptable.
If your own hydra can do it, you may set `meta.hydraPlatforms` in
your tests if you want that. (e.g. set it in a common module that's
imported in all your tests)

roberth commented:
I've moved the `darwin`-only definition back into the `runTest` default, which I didn't expect to be a feasible solution given the recent observations, but it seems that the `testing/meta.nix` solution doesn't have a certain flaw that the `addMetaAttrs` solution did. The `addMetaAttrs` call looks good on the face of it &ndash; I haven't got to the bottom of that, but I'm mostly just glad it works now.

vcunat commented:
I think that now these 2*5 jobs really aren't present (if I didn't make some typo).

roberth commented:
Anything I should do before we can merge this?

vcunat commented:
I'm not aware of anything.
